File: services/api-gateway/app/schemas/radiology.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ── Imaging Order schemas ─────────────────────────────────────────────────


class ImagingOrderCreate(BaseModel):
    """Schema for creating an imaging order."""

    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    modality: str = Field(
        ...,
        pattern=r"^(xray|ultrasound|ct|mri|fluoroscopy|mammography|ecg|echo)$",
    )
    body_part: str = Field(..., min_length=1, max_length=100)
    laterality: str | None = Field(
        None, pattern=r"^(left|right|bilateral|na)$"
    )
    views_requested: str | None = Field(None, max_length=200)
    study_description: str = Field(..., min_length=1, max_length=300)
    priority: str = Field(
        default="routine", pattern=r"^(stat|urgent|routine)$"
    )
    clinical_indication: str | None = Field(None, max_length=2000)
    clinical_history: str | None = Field(None, max_length=2000)
    contrast_required: bool = False
    contrast_type: str | None = Field(None, max_length=100)
    pregnancy_status: str | None = Field(
        None, pattern=r"^(not_pregnant|pregnant|unknown|na)$"
    )
    allergies_noted: str | None = Field(None, max_length=500)


class ImagingOrderResponse(BaseModel):
    """Schema for imaging order API responses."""

    id: uuid.UUID
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    ordered_by: uuid.UUID
    order_number: str
    accession_number: str | None
    modality: str
    body_part: str
    laterality: str | None
    views_requested: str | None
    study_description: str
    priority: str
    clinical_indication: str | None
    contrast_required: bool
    pregnancy_status: str | None
    status: str
    scheduled_at: datetime | None
    room: str | None
    performed_by: uuid.UUID | None
    performed_at: datetime | None
    total_cost_cents: int | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Worklist item (order + patient info) ──────────────────────────────────


class ImagingWorklistItem(BaseModel):
    """Imaging order in the worklist with patient details."""

    id: uuid.UUID
    order_number: str
    accession_number: str | None
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    ordered_by: uuid.UUID
    modality: str
    body_part: str
    laterality: str | None
    study_description: str
    priority: str
    status: str
    contrast_required: bool
    has_result: bool = False
    result_status: str | None = None
    is_critical: bool = False
    scheduled_at: datetime | None
    room: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ImagingWorklistResponse(BaseModel):
    """Imaging worklist response."""

    items: list[ImagingWorklistItem]
    total: int


# ── Schedule / Perform ────────────────────────────────────────────────────


class ImagingScheduleRequest(BaseModel):
    """Schema for scheduling an imaging study."""

    scheduled_at: datetime
    room: str | None = Field(None, max_length=50)


class ImagingPerformRequest(BaseModel):
    """Schema for marking an imaging study as performed."""

    accession_number: str | None = Field(None, max_length=50)
    notes: str | None = Field(None, max_length=2000)


# ── Imaging Result schemas ────────────────────────────────────────────────


class ImagingResultCreate(BaseModel):
    """Schema for creating/entering an imaging report (draft)."""

    findings: str | None = Field(None, max_length=10000)
    impression: str | None = Field(None, max_length=5000)
    recommendations: str | None = Field(None, max_length=5000)
    technique: str | None = Field(None, max_length=2000)
    comparison: str | None = Field(None, max_length=2000)
    urgency: str = Field(
        default="normal",
        pattern=r"^(normal|abnormal|critical|incidental)$",
    )
    is_critical: bool = False
    notes: str | None = Field(None, max_length=2000)


class ImagingResultResponse(BaseModel):
    """Schema for imaging result API responses."""

    id: uuid.UUID
    order_id: uuid.UUID
    patient_id: uuid.UUID
    reported_by: uuid.UUID | None
    verified_by: uuid.UUID | None
    findings: str | None
    impression: str | None
    recommendations: str | None
    technique: str | None
    comparison: str | None
    urgency: str
    is_critical: bool
    critical_notified: bool
    critical_notified_to: uuid.UUID | None
    critical_notified_at: datetime | None
    status: str
    reported_at: datetime | None
    verified_at: datetime | None
    image_keys: list[dict[str, str]] | None
    dicom_study_uid: str | None
    dicom_series_count: int | None
    dicom_instance_count: int | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ImagingResultVerify(BaseModel):
    """Schema for verifying (finalising) an imaging report."""

    notes: str | None = Field(None, max_length=2000)


class CriticalImagingNotifyRequest(BaseModel):
    """Schema for recording critical imaging finding notification."""

    notified_to: uuid.UUID


# ── Order Detail (order + result + patient) ───────────────────────────────


class ImagingOrderDetail(BaseModel):
    """Full imaging order with result and patient info."""

    order: ImagingOrderResponse
    result: ImagingResultResponse | None = None
    patient_name: str | None = None
    patient_mrn: str | None = None


# ── Radiology Summary ────────────────────────────────────────────────────


class RadiologySummary(BaseModel):
    """Dashboard summary stats for radiology department."""

    total_orders_today: int = 0
    pending_orders: int = 0
    in_progress: int = 0
    completed_today: int = 0
    pending_reports: int = 0
    critical_findings: int = 0

File: packages/shared/src/types/radiology.ts
/** Imaging modality types. */
export type ImagingModality =
  | "xray"
  | "ultrasound"
  | "ct"
  | "mri"
  | "fluoroscopy"
  | "mammography"
  | "ecg"
  | "echo";

/** Imaging order status. */
export type ImagingOrderStatus =
  | "ordered"
  | "scheduled"
  | "in_progress"
  | "completed"
  | "cancelled";

/** Imaging result/report status. */
export type ImagingResultStatus = "draft" | "preliminary" | "final" | "addendum";

/** Imaging urgency classification. */
export type ImagingUrgency = "normal" | "abnormal" | "critical" | "incidental";

/** Imaging order priority. */
export type ImagingPriority = "stat" | "urgent" | "routine";

/** Laterality. */
export type Laterality = "left" | "right" | "bilateral" | "na";

/** Pregnancy status for radiation safety. */
export type PregnancyStatus = "not_pregnant" | "pregnant" | "unknown" | "na";

/** Imaging order response from API. */
export interface ImagingOrderResponse {
  id: string;
  encounter_id: string;
  patient_id: string;
  ordered_by: string;
  order_number: string;
  accession_number: string | null;
  modality: ImagingModality;
  body_part: string;
  laterality: Laterality | null;
  views_requested: string | null;
  study_description: string;
  priority: ImagingPriority;
  clinical_indication: string | null;
  contrast_required: boolean;
  pregnancy_status: PregnancyStatus | null;
  status: ImagingOrderStatus;
  scheduled_at: string | null;
  room: string | null;
  performed_by: string | null;
  performed_at: string | null;
  total_cost_cents: number | null;
  created_at: string;
}

/** Imaging worklist item with patient info. */
export interface ImagingWorklistItem {
  id: string;
  order_number: string;
  accession_number: string | null;
  encounter_id: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  ordered_by: string;
  modality: ImagingModality;
  body_part: string;
  laterality: Laterality | null;
  study_description: string;
  priority: ImagingPriority;
  status: ImagingOrderStatus;
  contrast_required: boolean;
  has_result: boolean;
  result_status: ImagingResultStatus | null;
  is_critical: boolean;
  scheduled_at: string | null;
  room: string | null;
  created_at: string;
}

/** Imaging worklist response. */
export interface ImagingWorklistResponse {
  items: ImagingWorklistItem[];
  total: number;
}

/** Create imaging order request. */
export interface ImagingOrderCreate {
  encounter_id: string;
  patient_id: string;
  modality: ImagingModality;
  body_part: string;
  laterality?: Laterality | null;
  views_requested?: string | null;
  study_description: string;
  priority?: ImagingPriority;
  clinical_indication?: string | null;
  clinical_history?: string | null;
  contrast_required?: boolean;
  contrast_type?: string | null;
  pregnancy_status?: PregnancyStatus | null;
  allergies_noted?: string | null;
}

/** Schedule imaging request. */
export interface ImagingScheduleRequest {
  scheduled_at: string;
  room?: string | null;
}

/** Perform imaging study request. */
export interface ImagingPerformRequest {
  accession_number?: string | null;
  notes?: string | null;
}

/** Imaging result / report response. */
export interface ImagingResultResponse {
  id: string;
  order_id: string;
  patient_id: string;
  reported_by: string | null;
  verified_by: string | null;
  findings: string | null;
  impression: string | null;
  recommendations: string | null;
  technique: string | null;
  comparison: string | null;
  urgency: ImagingUrgency;
  is_critical: boolean;
  critical_notified: boolean;
  critical_notified_to: string | null;
  critical_notified_at: string | null;
  status: ImagingResultStatus;
  reported_at: string | null;
  verified_at: string | null;
  image_keys: Array<Record<string, string>> | null;
  dicom_study_uid: string | null;
  dicom_series_count: number | null;
  dicom_instance_count: number | null;
  notes: string | null;
  created_at: string;
}

/** Create / enter imaging report. */
export interface ImagingResultCreate {
  findings?: string | null;
  impression?: string | null;
  recommendations?: string | null;
  technique?: string | null;
  comparison?: string | null;
  urgency?: ImagingUrgency;
  is_critical?: boolean;
  notes?: string | null;
}

/** Verify imaging report. */
export interface ImagingResultVerify {
  notes?: string | null;
}

/** Critical imaging finding notification. */
export interface CriticalImagingNotifyRequest {
  notified_to: string;
}

/** Full imaging order detail with result. */
export interface ImagingOrderDetail {
  order: ImagingOrderResponse;
  result: ImagingResultResponse | null;
  patient_name: string | null;
  patient_mrn: string | null;
}

/** Radiology department summary stats. */
export interface RadiologySummary {
  total_orders_today: number;
  pending_orders: number;
  in_progress: number;
  completed_today: number;
  pending_reports: number;
  critical_findings: number;
}
